### PR TITLE
FixedResolutionBuffer image resolutions

### PR DESCRIPTION
**Pull request by ngoldbaum**
As noted in #1709, currently FixedResolutionBuffer will flip the image resolution for non-square buffers. The reason for this goes back to [Bitbucket pull request 2245](https://bitbucket.org/yt_analysis/yt/pull-requests/2245) where Matt rewrote how the pixelizers work. We actually had a similar problem even earlier than that that @yihaochen pointed out.

The basic issue is illustrated by this notebook:

https://gist.github.com/anonymous/4ab071eb7284f17ae062a2aee927fbcd

Matplotlib's imshow command interprets an image with shape (m, n) as m *rows* and n *columns*, which is backwards of yt's way of working with images where each pixel corresponds to an (x, y) location.

My proposed fix here is to make it so that if someone creates a FixedResolutionBuffer object manually or via a `to_frb` method hanging off a data object that we respect the resolution order specified by the user.

To maintain backwards compatibility for the way images look in `PlotWindow` plots and to comply with the natural user expectation that if I do `plot.set_buff_size((nx, ny))` there will be `nx` pixels along the x direction in the plot and `ny` pixels along the y direction, I've made it so that `set_buff_size` will reverse the order of the shape it passes to the `FixedResolutionBuffer` initializer. I've also added a note to the docstring of `set_buff_size` that this is the case to hopefully head off future confusion. I've also added comments throughout the codebase (or rewrote existing comments) to hopefully make it clearer to future readers of the codebase why we sometimes are reversing the image shape.

# tl;dr

Now when you create an FRB object directly, you get exactly the resolution you expect. The appearance of PlotWindow plots and the plot callbacks on a PlotWindow plot should be unaffected by this change.

### CODE REVIEW

**Comment by yihaochen** [comment]
This has become quite confusing. I am still not sure why this PR needs to change the sequence of buff_size. I would suggest maybe using different names when different conventions are used. For example, `buff_size` for yt convention (nx, ny), and `imshape` for matplotlib convention (nrow, ncol).

File: yt/visualization/plot_window.py [hunk 1]
```diff
@@ -578,11 +578,24 @@ def set_buff_size(self, size):
...
+            this way because matplotlib's imshow command displays images in
+            row-major order.
         """
         if iterable(size):
-            self.buff_size = size
+            # matplotlib displays images with shapes (m, n) with m rows
+            # and n columns. Reversing the shape here means that if I do
+            # plot.set_buff_size((nx, ny)) where will be nx pixels along
+            # x (nx columns) and ny pixels along y (ny rows).
+            self.buff_size = (size[1], size[0])
```
Comment (yihaochen) [inline yt/visualization/plot_window.py]: Why do you need to flip the size here? This will give a strange behavior that when I do `plot.set_buff_size((400, 800))`. I will get `plot.buff_size == ((800, 400))`

**Comment by ngoldbaum** [comment]
> I am still not sure why this PR needs to change the sequence of buff_size

If you're referring to my changes in the `PlotWindow.set_buff_size` function, it's to make sure that PlotWindow plots continue looking the same after this pull request. If I didn't make that change, then suddenly if someone does `plot.set_buff_size((800, 400))` they'd end up with an image that has 400 pixels along the x-axis of the plot and 800 pixels along the y-axis of the plot. Not only would that be a behavior change in the plotting system, I think it's confusing (even though it's what matplotlib does).

> For example, buff_size for yt convention (nx, ny), and imshape for matplotlib convention (nrow, ncol).

I'm not sure how I'd integrate your terminology into this pull request. Are you saying I should define that in the note I added to the `set_buff_size` docstring? I also don't think `imshape` is a good thing to define since matplotlib doesn't use it so we'd just be making up words to describe matplotlib's behavior. Do you find the wording of my note in the `set_buff_size` docstring confusing? Or the comments I added?

File: yt/visualization/plot_window.py (hunk 1 above)
Comment (ngoldbaum) [reply yt/visualization/plot_window.py]: Hmm, that's true. I guess I could do the flipping at a lower level.

Comment (ngoldbaum) [reply yt/visualization/plot_window.py]: Take a look at the latest version of the pull request, I think I've fixed this.

**Comment by ngoldbaum** [comment]
It looks like now the answer tests for the `ParticleProjectionPlot` tests are failing. Let me see if I can figure out a way to convince myself whether the current answers are correct or not. I suspect they are not and have been transposed for square image buffers and simply incorrect for rectangular image buffers.

**Comment by yihaochen** [comment]
> I'm not sure how I'd integrate your terminology into this pull request. Are you saying I should define that in the note I added to the set_buff_size docstring? I also don't think imshape is a good thing to define since matplotlib doesn't use it so we'd just be making up words to describe matplotlib's behavior. Do you find the wording of my note in the set_buff_size docstring confusing? Or the comments I added?

Never mind. I was somewhat thinking that you changed the meaning of buff_size so that buff_size = (ny, nx).

So after this PR the frb will have shape (nx, ny), which is different from the image array that has shape (ny, nx). I feel this contradicts with my previous understanding that frb **is** the image array.

This will have some other implications in the codes. For example, the `FITSImageData` takes an input of a `FixedResolutionBuffer`, a `YTCoveringGrid`, an `ImageArray`, or a `numpy.ndarray`. That will result in different orientations of the fits image depending on an ImageArray or an FRB is passed. (I did not check further, but passing a `YTCoveringGrid` to `FITSImageData` will probably also give a rotated image.)

**Comment by ngoldbaum** [comment]
> So after this PR the frb will have shape (nx, ny), which is different from the image array that has shape (ny, nx). I feel this contradicts with my previous understanding that frb is the image array.

That's not true, the FRB will generate images with shape `(nx, ny)` with this PR. Before this PR, yt had the exact issue you're worried that my PR would cause - the FRB generated images with a shape transposed from the shape specified by the `resolution` argument of the `FixedResolutionBuffer` initializer.

If you can make the concern you're having a bit more concrete (e.g. perhaps a small script explaining the issue?) that will help me to understand where you're coming from.

As far as the concern about FITSImageData, I don't have much experience with it (@jzuhone maybe you could weigh in here?) but here's an example of what I'm talking about:

```
import yt
ds = yt.load('IsolatedGalaxy/galaxy0030/galaxy0030')
slc = ds.slice(0, 0.0)
frb = slc.to_frb((500., "kpc"), (800, 400))
fid_frb = yt.FITSImageData(frb, fields=["density", "temperature"], units="pc")
print(fid_frb.shape)
```

Before this PR, this script prints `(400, 800)`, after it prints `(800, 400)`. This seems to me like with this PR yt is doing the correct thing and that the current version of yt is incorrectly flipping the resolution specified in the `to_frb` call.

Was there some other context where this PR might cause an incorrect result or change yt's behavior in a negative way?

**Comment by yihaochen** [comment]
It seems that I misunderstood the points of this PR. 
If one uses `set_buff_size((nx, ny))`, the frb will still have shape `(ny, nx)`. This PR will only affect directly calling `to_frb`. When one calls `to_frb(width, (nx, ny))`, the resulting frb array will have shape `(nx, ny)`. Is this correct?

**Comment by ngoldbaum** [comment]
Yup, that's correct. Before this PR the resulting image had shape `(ny, nx)` which is what prompted @phyzicist to open #1709.

**Comment by yihaochen** [comment]
So what I am concerned is that `to_frb` does not create an image-array-like array. I test with the following script:
```
import yt
import numpy as np
arr = np.arange(8).reshape(4,2,1)
data = dict(density = arr)
ds = yt.load_uniform_grid(data, arr.shape)

slc_plot = yt.SlicePlot(ds, 'z', 'density')
slc_plot.set_buff_size((4,2))
slc_plot.set_log('density', False)
slc_plot.show()
# This has shape (2,4)
print(slc_plot['density'].image._A.shape)

slc = ds.slice('z', 0.5)
slc_frb = slc.to_frb(1, (4,2))
# Currently, this gives an frb array of shape (2,4)
# [[ 0.  2.  4.  6.]
#  [ 1.  3.  5.  7.]]
#
# With PR 1710, this gives an frb array of shape (4,2), but averaged in wrong direction
# [[ 1.  5.]
#  [ 1.  5.]
#  [ 2.  6.]
#  [ 2.  6.]]
print(slc_frb['density'].shape)
print(slc_frb['density'])
```
I will work on an exmple for the FITSImageData later.

**Comment by ngoldbaum** [comment]
Yes, with this PR those two operations will create arrays that have different shapes. I'm arguing that the current behavior is broken and that we should change it so it is not broken anymore.

Both before and after this PR, `plot.set_buff_size((4, 2))` tells matplotlib to plot an image that has 4 columns and 2 rows. I think we both agree that this is the desired behavior.

Before this PR `slc.to_frb(1, (4, 2))` produced an image that had shape `(2, 4)`. From matpltolib's point of view this is the desired behavior (4 columns and 2 rows). However, from the yt users's point of view this is (in my opinion) backwards. For example, this behavior was confusing enough to @phyzicist that it prompted him to open an issue about it.  Why do the images I get out of my `FixedResolutionBuffer` object not have shape `(2, 4)` even though I specified `(4, 2)`?

With this PR, `slc.to_frb(1, (4, 2))` now produces an FRB that generates images with shape `(4, 2)`. As you note for the example in your script that isn't vert useful. But all you would need to do to fix that is say `slc.to_frb(1, (2, 4))` and then you'd get back the image you are expecting:

```
import yt
import numpy as np
arr = np.arange(8).reshape(4, 2, 1)
data = dict(density=arr)
ds = yt.load_uniform_grid(data, arr.shape)

slc = ds.slice('z', 0.5)
slc_frb = slc.to_frb(1, (2, 4))
print(slc_frb['density'].shape)
print(slc_frb['density'])
```

Which prints:

```
(2, 4)
[[0. 2. 4. 6.]
 [1. 3. 5. 7.]] g/cm**3
```

I don't see how the current behavior makes any sense. I do acknowledge that this is a backwards incompatible change and that people who are already working around this issue will need to update their scripts. I don't think backwards compatibility justifies keeping confusing behavior in the library.

**Comment by yihaochen** [comment]
I agree that the behavior of `set_buff_size` make sense.

I think there are two issues here.
First, after this PR, `plot.set_buff_size((4,2))` will create an internal FRB `plot[field].frb.shape == (2,4)`. However, if one calls `slc.to_frb(width, (4,2))`, it will give an FRB of shape (4,2). I found this confusing, but I can also understand that one would expect the returned FRB array should be consistent with the input resolution.

Second, even if with the proposed behavior, when I call `slc.to_frb(width, (4, 2))`, I would expect the returned array is the region divided into nx=4 and ny=2 pixels. So I would indeed expect to see the following array in my example:
```
[[ 0.  1.]
 [ 2.  3.]
 [ 4.  5.]
 [ 6.  7.]]
```
But with this PR, it seems that the region is divided into nx=2 and ny=4 pixels. I found this also confusing.

**Comment by ngoldbaum** [comment]
So this is an artifact of how yt orients images. If I do this, I get the array you're expecting:

```
import yt
import numpy as np
arr = np.arange(8).reshape(4, 2, 1)
data = dict(density=arr)
ds = yt.load_uniform_grid(data, arr.shape)

ds.coordinates.x_axis['z'] = 1
ds.coordinates.x_axis[2] = 1
ds.coordinates.y_axis['z'] = 0
ds.coordinates.y_axis[2] = 0
slc = ds.slice('z', 0.5)
slc_frb = slc.to_frb(1, (2, 4))
```

This prints:

```
[[0. 1.]
 [2. 3.]
 [4. 5.]
 [6. 7.]] g/cm**3
(4, 2)
```

So fundamentally nothing is wrong, you just want to make plots with the x-axis aligned vertically and the y-axis horizontal. This orientations happens to agree with the native ordering of the data, so yt doesn't end up needing to interpolate when it generates the FRB image.

With yt's default orientation, yt needs to do some interpolation to generate the FRB image. Since the resolution is low and the pixelization algorithms yt uses isn't really tuned to make nice results at poor resolution, you get a kinda junk result. 

That said, I don't think that means we should always return data with a shape opposite of what a user requests. We'd have the same problem if you were slicing along a different direction or wanted to call a different direction up.

**Comment by Yurlungur** [comment]
Hi guys, I'm here because of @ngoldbaum 's list email. Why not add a flag to `to_frb`, something like `natural`. So that `slc.to_frb(width, (4,2), natural=True)` returns what @yihaochen is expecting?

**Comment by phyzicist** [comment]
Wow, this is a lot more work than I imagined! Thanks for looking into it!

@ngoldbaum wrote:
>  I do acknowledge that this is a backwards incompatible change and that people who are already working around this issue will need to update their scripts. I don't think backwards compatibility justifies keeping confusing behavior in the library.

I would mention that I believe this indexing flipped between 3.3.x and 3.4.x, so that can of worms is already opened -- this has already caused backwards compatibility issues and script changes (which prompted my issue). I would be happy with either outcome if reasonable and documented!

@Yurlungur wrote:
> Why not add a flag to to_frb, something like natural. So that slc.to_frb(width, (4,2), natural=True) returns what @yihaochen is expecting

Seems very reasonable to enable both options! Although, what is "natural" for one person is not for another -- so maybe use a different variable name! NumPy addresses this problem with a flag "index={'ij', 'xy'}" in its meshgrid function:
https://docs.scipy.org/doc/numpy/reference/generated/numpy.meshgrid.html

**Comment by ngoldbaum** [comment]
> I would mention that I believe this indexing flipped between 3.3.x and 3.4.x, so that can of worms is already opened

This is true and despite all I've been arguing about up until now it might argue in favor of keeping the default the way it is and just improving the docstrings and documentation to make it clear what's going on, along with perhaps adding an additional keyword argument to the `to_frb` function to let users control which way they want the image indexed. Although that said, I suspect most users wouldn't notice the extra option.

**Comment by phyzicist** [comment]
My suggestion would be to revert to the 3.3.x behavior for defaults, since more scripts probably already exist that were developed in that version. But, I support the dev team whatever you decide.

**Comment by yihaochen** [comment]
I am copying my yt-dev message here.

I think the fundamental inconsistency comes from the way that an array is plotted in `matplotlib` (maybe it is just `imshow`?). In yt, I always index an array by `data[ix,iy]`. So the first index corresponds to x-axis, and the second index corresponds to y-axis. Therefore, if I specify a 2-d array with shape (4, 2), I expect the array to have 4 pixels in x-axis and 2 pixels in y-axis. I think this behavior is what most people would expect. However, if I pass the (4, 2) array into `imshow` (with `origin="lower"`), the first axis is actually plotted vertically, i.e. 4 pixels vertically and 2 pixels horizontally. In order to match the common sense of most people that x-axis (first axis) should be plotted horizontally, yt transposes the `ImageArray`, which comes from `FixedResolutionBuffer` that is also transposed.

Now we are trapped with two different conventions:

`raw_data[ix,iy]`

and

`FRB[iy,ix]`
`ImageArray[iy,ix]`

When a user wants to access deeper data like the `FRB`, different conventions will be encountered and likely cause confusions.

For example, if I specify
`size = (nx, ny)` - the argument in PlotWindow.set_buff_size
`resolution = (nx, ny)` - the argument in YTSelectionContainer2D.to_frb,
the resulting `FRB` and `ImageArray` will have the shape (ny, nx). This is what prompted the issue 1709 and this PR. And I agree that this is not a behavior that people would expect.

I feel that what we should really do is to unify all the conventions, letting `FRB` and the `ImageArray` both be indexed by `[ix, iy]` and have the shape `(nx, ny)`. Then, at the very end when the `ImageArray` is passed to imshow, we transpose it there (in `visualization/base_plot_types.py:ImagePlotMPL`). This would, of course, require a lot of work. But in my opinion, the unification would give us much more consistency throughout the yt code.

**Comment by ngoldbaum** [comment]
> This would, of course, require a lot of work.

In particular, it'll require touching the cython pizelizer code and partially backing out some of the changes Matt made in [Bitbucket PR 2245](https://bitbucket.org/yt_analysis/yt/pull-requests/2245).

That might be the best approach, I'll try to see if it's possible to do that in a sensible way.

**Comment by yihaochen** [comment]
I'd like to add to my previous mention of FITS data. It looks like that FITS files store the data in the way similar to `imshow`. For example in
https://python4astronomers.github.io/astropy/fits.html
```
>>> hdu.data.shape
(30, 360, 720)
```
The shape (30, 360, 720) corresponds to NAXIS1=720, NAXIS2=360, and NAXIS3=30 as seen in the header
```
NAXIS   =                    3 / number of data axes
NAXIS1  =                  720 / length of data axis 1
NAXIS2  =                  360 / length of data axis 2
NAXIS3  =                   30 / length of data axis 3
```
Using the current version of `FRB` or `ImageArray` will give correct orientation. However, I suspect that creating `FITSImageData` from `YTCoveringGrid` or raw data array will result in the wrong orientation.

**Comment by matthewturk** [comment]
@ngoldbaum Is this PR still valid?  It looks like it to me, but I wanted to check.

**Comment by ngoldbaum** [comment]
The concerns about backward compatibility would need to be addressed, someone would still need to make the pixelizers do the right thing.

**Comment by matthewturk** [comment]
@ngoldbaum Thanks!

**Comment by matthewturk** [comment]
@ngoldbaum I'll look at this, but will plan on closing if it doesn't seem tractable in the near future.
